api/envoy/api/v2/cds.proto

```diff
@@ -150,12 +151,26 @@ message Cluster {
   // when picking a host in the cluster.
   LbPolicy lb_policy = 6 [(validate.rules).enum.defined_only = true];
 
-  // If the service discovery type is
+  // Setting this is a requirement for clusters with discovery type
   // :ref:`STATIC<envoy_api_enum_value_Cluster.DiscoveryType.STATIC>`,
   // :ref:`STRICT_DNS<envoy_api_enum_value_Cluster.DiscoveryType.STRICT_DNS>`
-  // or :ref:`LOGICAL_DNS<envoy_api_enum_value_Cluster.DiscoveryType.LOGICAL_DNS>`,
-  // then hosts is required.
-  repeated core.Address hosts = 7;
+  // or :ref:`LOGICAL_DNS<envoy_api_enum_value_Cluster.DiscoveryType.LOGICAL_DNS>`.
+  // **This field is deprecated**. Set the :ref:`load_assignment<envoy_api_field_Cluster.load_assignment>`
+  // field instead.
+  repeated core.Address hosts = 7 [deprecated = true];
+
+  // Setting this is required for specifying members of
+  // :ref:`STATIC<envoy_api_enum_value_Cluster.DiscoveryType.STATIC>`,
```

Comment:
Can you update https://github.com/envoyproxy/envoy/blob/71c0fd695e992e21d2e0439eb14b449fa0b8aea4/api/envoy/api/v2/endpoint/endpoint.proto#L21 to indicate that the interpretation of the `Address` depends on the cluster type. For static, it will be expected to be a direct IP address (or something resolvable by the specified resolver in the `Address`). For logical/static DNS, it will be expected to be a hostname, resolved via DNS.

This is actually a bit more complicated thinking about it. See the wording at https://github.com/envoyproxy/envoy/blob/71c0fd695e992e21d2e0439eb14b449fa0b8aea4/api/envoy/api/v2/core/address.proto#L36. I think we should try and get clear in the API (with some implementation TODOs) what the final relationship is going to be between the custom resolvers (which can apply to listener addresses as well, and act on port values/names) and the DNS resolvers for `Cluster`s. @mattklein123 WDYT on this? Seems a rough edge in the API.

Reply:
Yeah I don't know. That's a good question. Should we just say that custom resolvers can't be used at all with DNS discovery types and block it? I don't think it really makes sense and is confusing. If we do that, I think we can have error checking / error messages to guide people in the right direction...